ads/google/utils.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {user} from '../../src/log';
+
+/**
+ * Given the amp-ad data attribute containing the multi-size dimensions, and a
+ * set of primary dimensions, this function will return all valid multi-size
+ * [width, height] pairs in an array.
+ *
+ * @param {string} multiSizeDataStr The amp-ad data attribute containing the
+ *   multi-size dimensions.
+ * @param {number} primaryWidth The primary width of the ad slot.
+ * @param {number} primaryHeight The primary height of the ad slot.
+ * @param {!Array<Array<number>>=} opt_dimensions An array into which to put
+ *   the multi-size dimensions.
+ * @param {boolean} multiSizeValidation A flag that if set to anything other
+ *   than false will enforce the rule that ensures multi-size dimensions are no
+ *   less than 2/3rds of their primary dimension counterpart.
+ * @return {!Array<Array<number>>} An array of arrays, the latter of which
+ *   contains the width and height as its sole elements (in that order).
+ */
+export function getMultiSizeDimensions(
+    multiSizeDataStr,
+    primaryWidth,
+    primaryHeight,
+    multiSizeValidation,
+    opt_dimensions) {
+
+  const dimensions = opt_dimensions || [];
+  const sizesStr = multiSizeDataStr.split(',');
+
+  sizesStr.forEach(sizeStr => {
+
+    const size = sizeStr.split('x');
+
+    // Make sure that each size is specified in the form val1xval2.
+    if (size.length != 2) {
+      user().error('AMP-AD', `Invalid multi-size data format '${sizeStr}'.`);
+      return;
+    }
+
+    const widthStr = size[0];
```

Comment:
cast these as numbers here, and then remove casting them below, changed the two validation arrow functions you pass in to validateDimensions so that it expects its already been converted with Number()